lib/puppet/provider/service/systemd.rb

```diff
@@ -129,13 +129,25 @@ def unmask
     systemctl_change_enable(:unmask)
   end
 
-  def restartcmd
-    [command(:systemctl), "restart", @resource[:name]]
+  def restart
+    cmds = []
+    cmds << :systemctl << 'restart' << @resource[:name]
+    begin
+      execute(cmds)
+    rescue Puppet::ExecutionFailure
+      raise Puppet::Error.new(prepare_error_message(@resource[:name], 'restart'), $!)
```

Comment:
Alternately, instead of using a magic variable, do this:

```ruby
rescue Puppet::ExecutionFailure => e
  raise Puppet::Error.new(prepare_error_message(@resource[:name], 'restart'), e)
```